refactor: Remove enforced sorted order of batch slots

Funatiq opened the pull request:
Change the order of decoder `batchSlots` by removing the sort [here](https://github.com/NVIDIA/TensorRT-LLM/pull/3502/files#r2041689369). 
This means that the order passed to the decoder is exactly the same as `seqSlots`. The only difference is that `seqSlots` contain all requests and `batchSlots` contain only finished context and generation requests.

This change simplifies the code and removes some extra steps and buffers. However, it introduces a sync after the decoder setup in the Eagle/ReDrafter/Lookahead case for correctness [here](https://github.com/NVIDIA/TensorRT-LLM/pull/3502#discussion_r2192403292). Without this the Eagle and ReDrafter tests might fail spuriously (so there is likely a race condition).

Funatiq committed:
refactor: Remove enforced sorted order of batch slots

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
test: Enable wavefront and random for spec decode variants

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
chore: Remove seqSlotRemapping

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
fixup: fix and simplify Eagle decoding

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run --stage-list "A10-CPP-1"

tensorrt-cicd commented:
[PR_Github #2048](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/2048/) [ run ] triggered by Bot

Funatiq commented on cpp/tensorrt_llm/runtime/runtimeKernels.cu:
```diff
@@ -1246,16 +1246,14 @@ void invokeUpdateKVBlockArrayDraftTokenLocation(ITensor const& seqAcceptedDraftT
-        bufferCast<SizeType32>(seqSlotRemapping), bufferCast<SizeType32>(batchSlots), maxKVCacheLen, maxBlocksPerSeq,
-        tokensPerBlock, canUseOneMoreBlock, stream);
+        sizeInBytesPerKVHead, rewindDraftTokenCommonCount, rewindDraftTokenSeparateAdjustments, nullptr,
+        bufferCast<SizeType32>(batchSlots), maxKVCacheLen, maxBlocksPerSeq, tokensPerBlock, canUseOneMoreBlock, stream);
```
It would be possible to remove the `seqSlotRemapping` parameter in the entire call stack.

nekorobov replied:
I think if we delete this, the logic with [this code ](https://github.com/NVIDIA/TensorRT-LLM/blob/main/cpp/tensorrt_llm/kernels/speculativeDecoding/kvCacheUpdateKernels.cu#L58) becomes wrong. We must access `pastKeyValueLengths`, which comes from `runtimeBuffers.sequenceLengthsDevice` according to the sorted `seqSlot` to sync the accesses between `runtimeBuffer` structs and `decoderBuffer` structs.

tensorrt-cicd commented:
[PR_Github #2048](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/2048/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #1507 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/1507/) completed with status: 'FAILURE'

Funatiq commented on cpp/tensorrt_llm/layers/eagleDecodingLayer.cpp:
```diff
@@ -278,27 +276,24 @@ void EagleDecodingLayer<T>::packAcceptedPaths(EagleOutputs const& outputs, Eagle
+    auto const* numNewTokens = bufferCast<SizeType32>(*outputs.numNewTokens.value());
+    auto* numNewTokensCumSum = bufferCast<SizeType32>(*outputs.numNewTokensCumSum);
+    auto* pathsOffsets = bufferCast<SizeType32>(*outputs.pathsOffsets);
+    auto const* batchSlots = augmentedSeqSlots;
```
We pass `augmentedSeqSlots` which means `seqSlots` with `-1` for skipped chunked requests.

nekorobov commented on cpp/tensorrt_llm/kernels/speculativeDecoding/common.cu:
```diff
@@ -81,22 +81,7 @@ __global__ void packAcceptedPaths(SizeType32* acceptedLengthsCumSum, SizeType32*
-                }
-            }
-        }
+        auto const pathBatchIdx = isPathsSeqSlotIdx ? bi : batchSlot;
```
Hm, sorry, I do not see the change related to that. Where do you change the batch slots order in `gptDecoderBatched`?

Funatiq replied:
In this kernel we use `augmentedSeqSlots` which means `seqSlots` with -1 for skipped chunked requests.

nekorobov replied:
Understood now, thanks!

Funatiq commented on cpp/tensorrt_llm/batch_manager/makeDecodingBatchInputOutput.cpp:
```diff
@@ -107,9 +107,6 @@ std::vector<SizeType32> getActiveSlots(RequestVector const& contextRequests, Req
         }
     }
-
-    std::sort(activeSlots.begin(), activeSlots.end());
```
Remove the sorting of decoder `batchSlots` here.

Funatiq commented:
/bot run --stage-list "A10-CPP-1"

tensorrt-cicd commented:
[PR_Github #5359](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/5359/) [ run ] triggered by Bot

Funatiq commented:
/bot run --stage-list "A10-CPP-1"

tensorrt-cicd commented:
[PR_Github #8898](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/8898/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #6483 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/6483/) completed with status: 'FAILURE'

Funatiq committed:
feat: Add NVTX profiling to BanWords, BeamSearch, and ExplicitDraftTokens layers

- Included NVTX3_SCOPED_RANGE for performance tracing in forwardAsync methods of BanWordsLayer, BeamSearchLayer, and ExplicitDraftTokensLayer.
- Added nvtxUtils.h header to the respective files for NVTX functionality.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
feat: Add NVTX profiling to decoding layers' setup methods for performance tracing

- Introduced NVTX3_SCOPED_RANGE for performance tracing in the setup methods of EagleDecodingLayer, ExternalDraftTokensLayer, LookaheadDecodingLayer, MedusaDecodingLayer, PenaltyLayer, SamplingLayer, TopKSamplingLayer, and TopPSamplingLayer.
- This enhancement aims to improve performance monitoring across various decoding layers.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq committed:
chore: Enhance logging for sequence slot management and request handling

- Updated error logging in `assignReqSeqSlots.cpp` and `sequenceSlotManager.cpp` to include request and sequence IDs for better traceability.
- Added warning logging in `trtGptModelInflightBatching.cpp` to indicate the iteration when no requests are scheduled due to insufficient resources.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

Funatiq commented:
/bot run --stage-list "A10-CPP-1"

Funatiq committed:
fix: Synchronize decoder setup to prevent race conditions

- Updated `trtGptModelInflightBatching.cpp` to synchronize the runtime stream after decoder setup to prevent race conditions.
- Added a comment explaining the synchronization to prevent future regressions.

Signed-off-by: Robin Kobus <19427718+Funatiq@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #11260](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11260/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #8328 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/8328/) completed with status: 'SUCCESS'

Funatiq commented on cpp/tensorrt_llm/batch_manager/trtGptModelInflightBatching.cpp:
```diff
@@ -1019,6 +1021,10 @@ void TrtGptModelInflightBatching::forwardAsync(RequestList const& activeRequests
+                // WAR: Sync to ensure that the decoder setup is complete before the context phase starts.
+                // Without this, there may be a race condition between the decoder setup and the context phase
+                // which also leads to spurious test failure in trtGptModelRealDecoderTest.
+                mRuntime->getStream().synchronize();
```
Not this WAR.

tensorrt-cicd commented:
[PR_Github #11301](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11301/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #8356 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/8356/) completed with status: 'SUCCESS'

Funatiq commented:
/bot run

tensorrt-cicd commented:
[PR_Github #11668](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11668/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #11668](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/11668/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #8640](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/8640/) completed with status: 'FAILURE'

Funatiq commented:
/bot run
